src/java.base/share/classes/jdk/internal/module/ModuleReferenceImpl.java

```diff
@@ -62,8 +62,12 @@ public class ModuleReferenceImpl extends ModuleReference {
     // ModuleResolution flags
     private final ModuleResolution moduleResolution;
 
-    // cached hash of this module to avoid needing to compute it many times
-    private byte[] cachedHash;
+    // Single-slot cache of this module's hash to avoid needing to compute
+    // it many times. For correctness under concurrent updates, we need to
+    // wrap the fields updated at the same time with a record, which carries
+    // the implicit final-field semantics for its members.
+    private record CachedHash(byte[] hash, String algorithm) {}
```

Comment:
The comment still seems a bit excessive, maybe drop the last bit "which carries ..." as it is not needed here.

Reply:
I don't think the comment is excessive. It declares the intent to  protects from future "optimization" of flattening the record into the plain fields. But I dropped "which carries".